### DISCUSSION TITLE
Handling optional features

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#598] [Dev Discussions]
Based on the discussion about precommit #356, I would like to discuss how we handle optional features.

**I think that our primary goal is the best out-of-the-box experience that puts in place all the best practices for Python development.**

However, no matter what decisions we make and how perfect we think it is, there will be users who need to modify the way pyscaffold works, by removing bits of the process, by adding customizations, or by replacing something.

I really hope we don't go down the "black" hole. Not saying that there is anything wrong with black, but there are a large number of developers that will prefer autopep8 because they can make it work they way they want.

If we generate the "perfect project" and also create a pyscaffold.conf file in the project (or in some other way record all the decisions about how it works) then the user can easily edit that "project configuration" and for example remove the precommit step.

I may already be opening up a can of worms, as I realize that ideally anything you can do in pyscaffold should be doable directly with the tool. So precommit is really part of the git configuration that we create.

In any case, can we have a general dicussion as opposed to just focusing on one instance of the issue/question?

**Comment by FlorianWilhelm** [comment]
Hi @julie777, that's an interesting dicussion. 

You perfectly stated the primary goal of PyScaffold. Maybe it helps us to look at some extremes:

1. **Perfect configurability & customization**: Let's imagine we would have a configuration file where you can choose from a lot of things, like the build systems (setuptools, poetry, flit, etc.), docs systems (Sphinx, mkdocs, ...), unit-testing framework, etc. 
 
   * Advantages:
    
      * users get exactly what they want if they know what they want
      * we would lose no more users, which for instance want mkdocs instead of Sphinx, and thus do not go with PyScaffold
    * Disadvantages: 
   
      * it's much harder for us to make clear what the best or recommended configuration is
      * maintaining PyScaffold would be much harder as with each option the number of configurations to test grows exponentially. For instance the build system, is like a fundament and a prerequisite for other tools like versioning (setuptools-scm). Allowing configurability here (some users asked for poetry instead of setuptools), would be a nightmare for us maintainers.  
      * PyScaffold's goal would overlap much more with the goals of Cookiecutter that allows you to define quite easily templates exactly matching your needs but of course then you have to also exactly know what you are doing. 

2. **No configurability at all**: If we would have no flags and no configurability, users would just call `putup my_proj` and we would select what we think it's best for them. 

   * Advantages:
    
      * PyScaffold's code complexity could be drastically decreased
      * less choices make the life of (inexperienced) developers easier
    * Disadvantages: 
   
      * We would need a "best-practice" for every aspect of PyScaffold. For CI/CD would this be Travis, Github, Gitlab or Cirrus? Since all of these have a strong community, it's not easy to go for a single one of them and thus we would dissatisfy users that are fine with 90% of what PyScaffold offers but want a specific CI/CD.


Looking at these extremes, I think it's clear that PyScaffold is somewhere in between. We want to provide users with some choices we made, like the one for the build systems, but want to allow configurability for others things like CI/CD. Maybe it would help us to cluster parts of a Python project setup and decide where we want to allow configurability and where not. Here is how I see our status quo right now but of course that's always for debate:

| Functionality | Configurability | Choice | Comment |
|---------------|-----------------|--------|---------|
|   build system     |       No         |     setuptools   | |
|  test/task runner | No | pytest with tox | some have asked for nox |
| code coverage | No | pytext-cov with coverage ||
| documentation | No | Sphinx with autodoc and RTM | fanbase of mkdocs seems to be growing |
| version control | No | git with setuptools-scm | every now and then someone asks for hg|
| artifact store | No | twine | from our docs ||
| environment manager | Yes | conda, virtualenv-based | has anyone ever tested how our project setup works with poetry's requirement management? I guess it doesn't| 
|   CI/CD             |    Yes             |  cirrus, gitlab, travis       |  Github in the future |
|   code quality           |      Yes           |  pre-commit -> black, isort, flake8, etc     |  some sane defaults are provided but are easily changed|

@abravalheri once came up with a smart extension system for PyScaffold and that helped us a lot to have configurability for users that want to change larger parts of PyScaffold without the need to touch the codebase of PyScaffold itself. This also allows us to test new extensions and eventually merge them with PyScaffold if they prove themselves, like we did it with the pyproject extension. 

@julie777, is this the direction you wanted to go with this discussion?

**Reply by julie777** [reply]
You have captured it well. 
The most important thing I got from your response is "what is our true goal?".

I spent significant time with Hatch before deciding to work on pyscaffold. Hatch does a very good job of providing full development process support.
![image](https://user-images.githubusercontent.com/5090005/153769899-3a546949-9c13-4f4e-9950-11fefb131693.png)
However, it really wants to control how users do things, "the one true way". It does have limited configuration. Much like black in being opinionated.

You also pointed out various part of the python development process and various tools and how in some cases we already support multiple tools. I really like the idea that we try to support the "best" tool by default, but the best tool will change over time. I found your table interesting, because before I started participating I created a huge spreadsheet of all the python dev process tools that I could find and categorized them by the parts of the process that they supported. (I'll add a separate entry about that.)

Your mention of Poetry and CookieCutter are great examples. 

I found that when I looked at Poetry it supports multiple orthogonal parts of the dev process. I think this makes it harder to integrate with other tools. However, I didn't spend a large amount of time getting to know it completely.

As for cookiecutter, I look at both it and copier and found that they are very general purpose and much more complicated that I would like, and that they actually do what I see as the primary built-in functionality of pyscaffold.

**Comment by julie777** [comment]
This is really starting to overlap with the other discussion that I started #592. Now that I have immersed myself in the code and gotten past the wow of pyscaffold having such great docs, I am seeing some things that I don't understand.

I will give you a couple of examples and then back off to see where the discussion goes.

1. A python best practice is to use a virtual environment for development. But, `putup` doesn't create one by default. It creates dummy test files, and it supports tox creating environment to run pytest, but it doesn't set me up to immediately begin coding. I expect to write some code and then run pytest. Repeat. I won't use tox until I want to run a full test in multiple environments. I have been using virtualenv and virtualenvwrapper to create a dev virtualenv and use it to iterate the development loop (write a test, run it, write some code, run the test...). This is completely independent from running tox to "test all" and do packaging, etc.

2. `putup` doesn't run `pip -e .` to actually turn the project into an installed working project.

3. `putup` doesn't provide a way for me to say "for every project I create I want pylint, behave, boltons... installed" and it doen't install the dependencies automatically into the virtual env that (it didn't create) that i will use for dev.

What I see right now is that Hatch provides a complete user experience and pyscaffold isn't there yet.
```
$ hatch -h
Usage: hatch [OPTIONS] COMMAND [ARGS]...

Options:
  --version   Show the version and exit.
  -h, --help  Show this message and exit.

Commands:
  build      Builds a project
  clean      Removes a project's build artifacts
  conda      Installs Miniconda
  config     Locates, updates, or restores the config file
  env        Manages virtual environments
  grow       Increments a project's version
  init       Creates a new Python project in the current directory
  install    Installs packages
  new        Creates a new Python project
  pypath     Manages named Python paths
  python     Manages Python installations
  release    Uploads to PyPI
  shed       Removes named Python paths or virtual environments
  shell      Activates or sends a command to a virtual environment
  test       Runs tests
  uninstall  Uninstalls packages
  update     Updates packages
```
But when I read the pypa discussion about pyscaffold I saw the potential to do everything that hatch does, and more. PyScaffold can be the standard dev process tool and support exploring new individual tools withing the process scaffold, provide the current best tools by default, etc. per #592. However, I don't know what you (the owners and maintainers of pyscaffold) believe or even want pyscaffold to be.

**Reply by abravalheri** [reply]
> 1. A python best practice is to use a virtual environment for development. But, `putup` doesn't create one by default. It creates dummy test files, and it supports tox creating environment to run pytest, but it doesn't set me up to immediately begin coding. I expect to write some code and then run pytest. Repeat. I won't use tox until I want to run a full test in multiple environments. I have been using virtualenv and virtualenvwrapper to create a dev virtualenv and use it to iterate the development loop (write a test, run it, write some code, run the test...). This is completely independent from running tox to "test all" and do packaging, etc.

Please note that you can run `putup --venv` to create a virtual environment. This is not enforced because users can have multiple ways of doing it (e.g. datascience people might decide to use `conda`). 

Something to also be noted is that users can pass any argument directly to `pytest` with `tox -- <pytest args>`. Personally I use this as lot to restrict the amount of tests running and focus on specific features/bugs.

> 2. putup doesn't run pip -e . to actually turn the project into an installed working project.

Currently the project is not installed when you run `putup --venv` but that could be improved? (see https://github.com/pyscaffold/pyscaffold/discussions/601#discussioncomment-2213773)

**Reply by abravalheri** [reply]
> 3\. `putup` doesn't provide a way for me to say "for every project I create I want pylint, behave, boltons... installed"

Assuming the relevant extensions exist, you can add that to your `~/.config/pyscaffold/default.cfg` (see [configuring docs](https://pyscaffold.org/en/stable/usage.html#pyscaffold-s-own-configuration)).

That file can be easily created by creating a project and passing the `--save-config` flag.

**Reply by julie777** [reply]
> Please note that you can run `putup --venv` to create a virtual environment. This is not enforced because users can have multiple ways of doing it (e.g. datascience people might decide to use `conda`).

I knew about --venv, but again, it isn't the default and it is provided by an extension.
> 
> Something to also be noted is that users can pass any argument directly to `pytest` with `tox -- <pytest args>`. Personally I use this as lot to restrict the amount of tests running and focus on specific features/bugs.

I also new about this, but running with tox means that it has to create environments and update them. Then I have to tell it which one to use, because for quick turn-around while coding I just want to run fast tests in my dev environment. You think of it as unifying, while I think of it as a wrapper around pytest that causes me to think more and type more. 

Part of my issues with tox is the documentation. I can't comment about the sufficiency of the content, but the organization is terrible. It starts of with a strange combination of single column and multi-column choices in the sidebar. Then other pages have their own menu (links to sections). I still haven't found cli documentation. And from `tox -h` help I don't understand how to specify a single environment to run tests in.  (For the record, I am struggling with tox in some other project also. Mostly because they aren't really setup to run on Windows.)

> 
> > 2. putup doesn't run pip -e . to actually turn the project into an installed working project.
> 
> Currently the project is not installed when you run `putup --venv` but that could be improved? (see [#601 (comment)](https://github.com/pyscaffold/pyscaffold/discussions/601#discussioncomment-2213773))

I will look at it. Much of this is just an issue of what our defaults are.

**Comment by FlorianWilhelm** [comment]
Could we please try to stick to the actual subject of this discussion? I feel like everything is quite mixed and I am having a hard time following what we are discussing right now.

**Reply by julie777** [reply]
sorry. Fixed.
